drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,502 @@
+	    ((time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) < -128 ||
+	     (time_value->tm_year - RTC_NXP_IRTC_YEAR_OFFSET) > 127)) {
+		return 0;
+	}
+
+	if ((RTC_NXP_RTCTIME_COMPARE_FLAG(mask, RTC_ALARM_TIME_MASK_WEEKDAY)) &&
+	    (time_value->tm_wday < -1 || time_value->tm_wday > 6)) {
+		return 0;
+	}
+
+	if ((mask == 0) && (time_value->tm_isdst < -1 || time_value->tm_isdst > 1)) {
+		return 0;
+	}
+
+	return 1;
+}
+
+/*
+ * The runtime where this is accessed is unknown
+ * so we force a lock on the registers
+ * then force an unlock to guarantee
+ * 2 seconds of write time
+ */
+static void nxp_irtc_unlock_registers(RTC_Type *reg)
+{
+	/* Lock the regsiters */
+	while ((reg->STATUS & (uint16_t)RTC_STATUS_WRITE_PROT_EN_MASK) == 0) {
+		*(uint8_t *)(&reg->STATUS) |= RTC_STATUS_WE(0x2);
+	}
+	/* Unlock the registers */
+	while ((reg->STATUS & (int16_t)RTC_STATUS_WRITE_PROT_EN_MASK) != 0) {
+		/*
+		 * The casting is required to update the
+		 * field of this register, the pattern here
+		 * unlocks the RTC registers for writing.
+		 * When unlocked if a 0x20 if written to this register
+		 * the registers will lock again and the next write
+		 * will lead to a fault.
+		 */
+		*(volatile int8_t *)(&reg->STATUS) = 0x00;
+		*(volatile int8_t *)(&reg->STATUS) = 0x40;
+		*(volatile int8_t *)(&reg->STATUS) = 0xC0;
+		*(volatile int8_t *)(&reg->STATUS) = 0x80;
+	}
+}
+
+static int nxp_irtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr) {
+		return -ENODATA;
+	}
+	if (!is_rtc_time_in_bounds(timeptr, 0)) {
+		return -EINVAL;
+	}
+
+	int calc_year = timeptr->tm_year - RTC_NXP_IRTC_YEAR_OFFSET;
+	/* The IRTC Month Index starts at 1 instead of 0 */
+	int calc_month = timeptr->tm_mon + 1;
+
+	uint32_t key = irq_lock();
+
+	nxp_irtc_unlock_registers(irtc_reg);
+	irtc_reg->SECONDS = RTC_SECONDS_SEC_CNT(timeptr->tm_sec);
+
+	uint32_t bits_to_set =
+		RTC_HOURMIN_MIN_CNT(timeptr->tm_min) | RTC_HOURMIN_HOUR_CNT(timeptr->tm_hour);
+	irtc_reg->HOURMIN = bits_to_set;
+
+	/*
+	 * -1 is valid for rtc_time.tm_wday property but is out of
+	 * bounds for IRTC registers
+	 */
+	bits_to_set = 0;
+	bits_to_set = RTC_DAYS_DAY_CNT(timeptr->tm_mday) |
+		      (timeptr->tm_wday == -1 ? 0 : RTC_DAYS_DOW(timeptr->tm_wday));
+	irtc_reg->DAYS = bits_to_set;
+
+	bits_to_set = 0;
+	bits_to_set = RTC_YEARMON_MON_CNT(calc_month) | RTC_YEARMON_YROFST(calc_year);
+	irtc_reg->YEARMON = bits_to_set;
+
+	if (timeptr->tm_isdst != -1) {
+		irtc_reg->CTRL |= RTC_CTRL_DST_EN(timeptr->tm_isdst);
+		data->is_dst_enabled = true;
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int nxp_irtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	struct nxp_irtc_data *data = dev->data;
+	RTC_Type *irtc_reg = config->base;
+
+	if (!timeptr) {
+		return -ENODATA;
```

Comment:
same comment/question as above function

Reply:
You need to be more descriptive with this one, I do not see what comment this is referring to?